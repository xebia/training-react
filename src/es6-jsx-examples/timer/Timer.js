import React, { Component } from 'react';

function getState() {
  return {
    time: new Date(),
  };
}

class Timer extends Component {
  constructor(props) {
    super(props);
    this.state = getState();
  }

  componentWillMount() {
    const updateState = () => this.setState(getState());
    this.interval = setInterval(updateState, 500);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  render() {
    return (<div>
      <h1>Hello World!</h1>
      <p>The time is {this.state.time.toString()}</p>
    </div>);
  }
}

export default Timer;
