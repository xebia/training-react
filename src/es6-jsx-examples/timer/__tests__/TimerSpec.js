import React from 'react';
import { shallow } from 'enzyme';
import Timer from '../Timer';

jest.useFakeTimers();
Date = jest.fn();
Date.now = jest.fn();

describe('Component: Timer', () => {
  let wrapper;

  beforeEach(() => {
    jest.clearAllMocks();
    setInterval.mockRestore();
    wrapper = shallow(<Timer />);
  });
  
  afterEach(() => {
    jest.clearAllTimers();
  });

  it('should render the current time based on the state', () => {
    wrapper.setState({ time: new Date(10000) });
    expect(wrapper).toMatchSnapshot();
  });

  it('should update state.time every 500ms', () => {
    expect(wrapper.state().time).toBe(Date.mock.instances[0]);
    jest.runTimersToTime(499);
    expect(wrapper.state().time).toBe(Date.mock.instances[0]);
    jest.runTimersToTime(1);
    expect(wrapper.state().time).toBe(Date.mock.instances[1]);
    jest.runTimersToTime(500);
    expect(wrapper.state().time).toBe(Date.mock.instances[2]);
  });

  it('should clear the interval on unmount', () => {
    expect(clearInterval).not.toHaveBeenCalled();
    wrapper.unmount();
    expect(clearInterval).toHaveBeenCalled();
  });
});
