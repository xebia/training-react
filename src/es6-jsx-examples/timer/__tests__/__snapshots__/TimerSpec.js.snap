// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Timer should render the current time based on the state 1`] = `
<div>
  <h1>
    Hello World!
  </h1>
  <p>
    The time is 
    [object Object]
  </p>
</div>
`;
